# Youtube comments analysis
In this notebook, we have a dataset of user comments for youtube videos related to animals or pets. We will attempt to identify cat or dog owners based on these comments, find out the topics important to them, and then identify video creators with the most viewers that are cat or dog owners.

The dataset are comments for videos related to animals and/or pets and is 240MB compressed.

In [3]:
link: https://drive.google.com/file/d/1o3DsS3jN_t2Mw3TsV0i7ySRmh9kyYi1a/view?usp=sharing
command: wget https://drive.google.com/file/d/1o3DsS3jN_t2Mw3TsV0i7ySRmh9kyYi1a/view?usp=sharing

## Data Exploration and Cleaning

In [5]:
df_unclean = spark.read.csv("/FileStore/tables/animals_comments.csv",inferSchema=True,header=True)
df_unclean.show(10)

+--------------------+------+-------------------------------------+
 creator_name|userid| comment|
+--------------------+------+-------------------------------------+
 Doug The Pug| 87.0| I shared this to ...|
 Doug The Pug| 87.0| Super cute 😀🐕🐶|
 bulletproof| 530.0| stop saying get e...|
 Meu Zoológico| 670.0| Tenho uma jiboia ...|
 ojatro|1031.0| I wanna see what ...|
 Tingle Triggers|1212.0| Well shit now Im ...|
Hope For Paws - O...|1806.0| when I saw the en...|
Hope For Paws - O...|2036.0| Holy crap. That i...|
 Life Story|2637.0|武器はクエストで貰えるんじゃないん...|
 Brian Barczyk|2698.0| Call the teddy Larry|
+--------------------+------+-------------------------------------+
only showing top 10 rows

In [6]:
df_unclean.count() 

Out[ 2 ]: 5820035

In [7]:
from pyspark.sql.functions import isnan, when, count, col, isnull

df_unclean.select([count(when(isnull('comment'), 'comment')).alias('commentIsNull')]).show()
#df_unclean.where(col("comment").isNull()).count()

+-------------+
commentIsNull|
+-------------+
 1051|
+-------------+

In [8]:
df_clean = df_unclean.na.drop(subset=["comment"])
df_clean.count()

Out[ 3 ]: 5818984

In [9]:
df_clean.show()

+--------------------+------+-------------------------------------+
 creator_name|userid| comment|
+--------------------+------+-------------------------------------+
 Doug The Pug| 87.0| I shared this to ...|
 Doug The Pug| 87.0| Super cute 😀🐕🐶|
 bulletproof| 530.0| stop saying get e...|
 Meu Zoológico| 670.0| Tenho uma jiboia ...|
 ojatro|1031.0| I wanna see what ...|
 Tingle Triggers|1212.0| Well shit now Im ...|
Hope For Paws - O...|1806.0| when I saw the en...|
Hope For Paws - O...|2036.0| Holy crap. That i...|
 Life Story|2637.0|武器はクエストで貰えるんじゃないん...|
 Brian Barczyk|2698.0| Call the teddy Larry|
 The Dodo|2702.0| 😐🤔😓😢😭😭😭😭😟|
Hope For Paws - O...|2911.0| That mother cat l...|
Hope For Paws - O...|2911.0| Its people like H...|
 Talking Kitty Cat|2911.0| steve: No wet foo...|
 Brave Wilderness|3224.0| Dont call this a ...|
 MaxluvsMya|3267.0| why are you alway...|
Rise Up Society F...|3372.0| Deb Tucker|
 The Dodo|3466.0| Thats a deer isnt...|
 Brave Wilderness|3466.0| there is no safe ...|
 Brave Wilderness|3466.0| Red before yellow...|
+--------------------+------+-------------------------------------+
only showing top 20 rows

In [10]:
display(df_clean)

creator_name userid comment Doug The Pug 87.0 I shared this to my friends and mom the were lol Doug The Pug 87.0 Super cute 😀🐕🐶 bulletproof 530.0 stop saying get em youre literally dumb . have some common sense or dont own this kind of dog. fucking retarded I swear Meu Zoológico 670.0 Tenho uma jiboia e um largato ojatro 1031.0 I wanna see what happened to the pigs after that please Tingle Triggers 1212.0 Well shit now Im hungry Hope For Paws - Official Rescue Channel 1806.0 when I saw the end it said to adopt I saw different animal sites I was mad that they separated the cute little pups after being together for a long time Hope For Paws - Official Rescue Channel 2036.0 Holy crap. That is quite literally the most adorable pup Ive ever seen. Life Story 2637.0 武器はクエストで貰えるんじゃないんですか？ Brian Barczyk 2698.0 Call the teddy Larry The Dodo 2702.0 😐🤔😓😢😭😭😭😭😟 Hope For Paws - Official Rescue Channel 2911.0 That mother cat looks like my own Im guessing she is a russian blue due to her looks and unusual coping skills. Hope For Paws - Official Rescue Channel 2911.0 Its people like Hope For Paws who truly make the world a better place <3 Talking Kitty Cat 2911.0 steve: No wet food for a month!:cats immediately stop fighting: Brave Wilderness 3224.0 Dont call this a challenge because there will be idiots who actually try this MaxluvsMya 3267.0 why are you always awake so late lol Rise Up Society Fan Page 3372.0 Deb Tucker The Dodo 3466.0 Thats a deer isnt it? Brave Wilderness 3466.0 there is no safe way to hold a crab. Somehow they will always manage to nip you! Brave Wilderness 3466.0 Red before yellow will kill a fellow...Some of you will get this ;) Brian Barczyk 3589.0 Im 7 and I love your videosAnd love your snakes Info Marvel 4504.0 Falto deadpool en el trairler Gohan The Husky 4533.0 2:35 Why do you leave the TV on if you are not at home?...... Brave Wilderness 4900.0 No its because of me. JK :D Congrats Vet Ranch 5005.0 can he hear ? Taylor Nicole Dean 5697.0 amazing family you have Vet Ranch 5960.0 29 likes and 3 views?? TOMMYfromtheBRONX 6835.0 Good on ya Tommy boy!! GoHerping 7841.0 Can I call you meme daddy GoHerping 7841.0 your becoming a meme channel I love it Brave Wilderness 8786.0 WOW I got dooped Kitsune Kreations 8857.0 I will not put lyrics and I’m glad that heather is bi ^^Would backup be possible? Angel Hawk 8857.0 Is Vural a Serval or a hyena? I’m just curious Kitsune Kreations 8857.0 Part 36 Graystripe X Firestar and if you call it a crack ship I will make you listen to my singing The Dodo 8935.0 😱😻 ViralHog 9312.0 So pure ❤️❤️❤️❤️ Abbey Normal 9312.0 did the animal come to you and ask to be sacrificed? No. The animal didnt give itself to you. Nathan Coffee 9431.0 Its so cute!! HoppingHammy 9431.0 its ok girl we would rather wait for more vids and you feeling better than the opposite :) Taylor Nicole Dean 9431.0 omg you can see casey simpson in the bottom right corner at 4:57 Cole & Marmalade 9431.0 cat drugs Cole & Marmalade 9431.0 with the green screen you should somehow do cats in space! :P Viktor Larkhill 10269.0 hey victor..you are wonderful human being..have no words to express the beautiful work that you are going.. you have inspired me in a way..I wish you and all your dogs great health and long long life.. love from India ! Taylor Nicole Dean 10336.0 I dont understand how you think she will make a good service dogs. SD are handled by a company here in Quebec (and given for free fully trained to people in need). For the first year they are fostered by families who expose them to as many things as possible and even then after a year the majority of them are deemed not fit for service work. They have to be ultra confident and never startled never afraid of anything etc... She seem like a good pet but the very opposite of what a service dog should be. Keedes channel LIVE 10683.0 Аа Keedes channel LIVE 10683.0 Я хочу зобрать Рыжыка Brian Barczyk 10749.0 Do your boss let you do this Rachel Fusaro 10765.0 Im not allowe

In [11]:
# find user with preference of dog and cat
from pyspark.sql.functions import isnan, when, count, col, isnull

df_clean = df_clean.withColumn("label", \
                           (when(col("comment").like("%my dog%"), 1) \
                           .when(col("comment").like("%I have a dog%"), 1) \
                           .when(col("comment").like("%my cat%"), 1) \
                           .when(col("comment").like("%I have a cat%"), 1) \
                           .when(col("comment").like("%my puppy%"), 1) \
                           .when(col("comment").like("%my pup%"), 1) \
                           .when(col("comment").like("%my kitty%"), 1) \
                           .when(col("comment").like("%my pussy%"), 1) \
                           .otherwise(0)))

In [12]:
df_clean.show()

+--------------------+------+-------------------------------------+-----+
 creator_name|userid| comment|label|
+--------------------+------+-------------------------------------+-----+
 Doug The Pug| 87.0| I shared this to ...| 0|
 Doug The Pug| 87.0| Super cute 😀🐕🐶| 0|
 bulletproof| 530.0| stop saying get e...| 0|
 Meu Zoológico| 670.0| Tenho uma jiboia ...| 0|
 ojatro|1031.0| I wanna see what ...| 0|
 Tingle Triggers|1212.0| Well shit now Im ...| 0|
Hope For Paws - O...|1806.0| when I saw the en...| 0|
Hope For Paws - O...|2036.0| Holy crap. That i...| 0|
 Life Story|2637.0|武器はクエストで貰えるんじゃないん...| 0|
 Brian Barczyk|2698.0| Call the teddy Larry| 0|
 The Dodo|2702.0| 😐🤔😓😢😭😭😭😭😟| 0|
Hope For Paws - O...|2911.0| That mother cat l...| 0|
Hope For Paws - O...|2911.0| Its people like H...| 0|
 Talking Kitty Cat|2911.0| steve: No wet foo...| 0|
 Brave Wilderness|3224.0| Dont call this a ...| 0|
 MaxluvsMya|3267.0| why are you alway...| 0|
Rise Up Society F...|3372.0| Deb Tucker| 0|
 The Dodo|3466.0| Thats a deer isnt...| 0|
 Brave Wilderness|3466.0| there is no safe ...| 0|
 Brave Wilderness|3466.0| Red before yellow...| 0|
+--------------------+------+-------------------------------------+-----+
only showing top 20 rows

In [13]:
print("The number of data with label 1:", df_clean.filter(col('label') == 1).count())
print("The number of data with label 0:", df_clean.filter(col('label') == 0).count())

The number of data with label 1: 40225
The number of data with label 0: 5778759

Notice that the label is very unbalanced in the orginal dataset and the pipeline transformation is pretty slow on the whole dataset. So here I took 70% of the data with label 1 and 5% of the data with label 0 to do further analysis.

In [15]:
df_clean_tmp1 = df_clean.filter(col('label') == 1).randomSplit([0.7, 0.3], seed = 100)
df_clean_tmp0 = df_clean.filter(col('label') == 0).randomSplit([0.005, 0.995], seed = 100)

df_clean_sub = df_clean_tmp1[0].union(df_clean_tmp0[0])

In [16]:
df_clean_sub.count()

Out[ 26 ]: 56905

The new sample data has 56905 observations.

## Data Preprocessing

In [19]:
from pyspark.ml.feature import RegexTokenizer, Word2Vec

# regular expression tokenizer
# Python 3: "word character": Unicode letter, ideogram, digit, or underscore
regexTokenizer = RegexTokenizer(inputCol="comment", outputCol="words", pattern="\\W")
word2Vec = Word2Vec(inputCol="words", outputCol="features")

In [20]:
from pyspark.ml import Pipeline

pipeline = Pipeline(stages=[regexTokenizer, word2Vec])

# Fit the pipeline to training documents.
pipelineFit = pipeline.fit(df_clean_sub)
dataset = pipelineFit.transform(df_clean_sub)

In [21]:
#(lable0_train,lable0_test) = dataset.filter(col('label') == 1).randomSplit([0.7, 0.3],seed = 100)
#(lable1_train, lable1_ex) = dataset.filter(col('label') == 0).randomSplit([0.005, 0.995],seed = 100)
#(lable1_test, lable1_ex2) = lable1_ex.randomSplit([0.002, 0.998],seed = 100)

(trainingData, testData) = dataset.randomSplit([0.7, 0.3], seed = 100)

Confirmed that the training data and test data has balanced labels.

In [23]:
print("Dataset Count: " + str(dataset.count()))
print("Training Dataset Count: " + str(trainingData.count()))
print("Test Dataset Count: " + str(testData.count()))
print("Training Dataset with Label 1 Count: " + str(trainingData.filter(col('label') == 1).count()))
print("Training Dataset with Label 0 Count: " + str(trainingData.filter(col('label') == 0).count()))
print("Test Dataset with Label 1 Count: " + str(testData.filter(col('label') == 1).count()))
print("Test Dataset with Label 1 Count: " + str(testData.filter(col('label') == 0).count()))

Dataset Count: 56905
Training Dataset Count: 39958
Test Dataset Count: 16947
Training Dataset with Label 1 Count: 19910
Training Dataset with Label 0 Count: 20048
Test Dataset with Label 1 Count: 8378
Test Dataset with Label 1 Count: 8569

## Models

### LogisticRegression

In [26]:
from pyspark.ml.classification import LogisticRegression, RandomForestClassifier, GBTClassifier
from pyspark.ml.evaluation import BinaryClassificationEvaluator
from pyspark.ml.tuning import ParamGridBuilder, CrossValidator
evaluator = BinaryClassificationEvaluator(rawPredictionCol = "rawPrediction", labelCol = "label")

In [27]:
lr = LogisticRegression(labelCol = "label", featuresCol = "features", maxIter = 10)

model = lr.fit(trainingData)

prediction_lr_train = model.transform(trainingData)
prediction_lr = model.transform(testData)

In [28]:
evaluator = BinaryClassificationEvaluator(rawPredictionCol = "rawPrediction", labelCol = "label")

print("The area under ROC for train set with logistic regression before CV is {}".format(evaluator.evaluate(prediction_lr_train)))
print("The area under ROC for test set with logistic regression before CV is {}".format(evaluator.evaluate(prediction_lr)))

The area under ROC for train set with logistic regression before CV is 0.9472373385742665
The area under ROC for test set with logistic regression before CV is 0.946305851749104

#### Parameter Tuning and K-fold cross-validation

In [30]:
paramGrid = (ParamGridBuilder()
             .addGrid(lr.regParam, [0.01, 0.5, 2.0])
             .addGrid(lr.elasticNetParam, [0.0, 0.5, 1.0])
             .addGrid(lr.maxIter, [5, 10, 15])
             .build())

cv = CrossValidator(estimator = lr, estimatorParamMaps = paramGrid, evaluator = evaluator, numFolds = 5)

# Run cross validations
cvModel = cv.fit(trainingData)
prediction_lr_best = cvModel.transform(testData)

print("The area under ROC for test set with logistic regression after CV is {}".format(evaluator.evaluate(prediction_lr_best)))

The area under ROC for test set with logistic regression after CV is 0.9515137966579144

### RandomForest

In [32]:
# Create an initial RandomForest model.
rf = RandomForestClassifier(labelCol="label", featuresCol="features")

# Train model with Training Data
rfModel = rf.fit(trainingData)

# Use test set here so we can measure the accuracy of our model on new data
prediction_rf = rfModel.transform(testData)

# cvModel uses the best model found from the Cross Validation
# Evaluate best model
print("The area under ROC for test set with random forest before CV is {}".format(evaluator.evaluate(prediction_rf)))

The area under ROC for test set with random forest before CV is 0.925232593652787

#### Parameter Tuning and K-fold cross-validation

In [34]:
rf = RandomForestClassifier(labelCol="label", featuresCol="features")

paramGrid = (ParamGridBuilder()
             .addGrid(rf.maxDepth, [2, 4, 6])
             .addGrid(rf.maxBins, [20, 40, 60])
             .addGrid(rf.numTrees, [5, 15, 25])
             .build())

# Create 5-fold CrossValidator
cv = CrossValidator(estimator = rf, estimatorParamMaps = paramGrid, evaluator = evaluator, numFolds = 5)

# Run cross validations.  This can take about 6 minutes since it is training over 20 trees!
cvModel = cv.fit(trainingData)

# Use test set here so we can measure the accuracy of our model on new data
prediction_rf_best = cvModel.transform(testData)

# cvModel uses the best model found from the Cross Validation
# Evaluate best model
print("The area under ROC for test set with random forest after CV is {}".format(evaluator.evaluate(prediction_rf_best)))

The area under ROC for test set with random forest after CV is 0.9252728075612519

### Gradient boosting

In [36]:
# Create an initial RandomForest model.
gbt = GBTClassifier(labelCol="label", featuresCol="features")

# Train model with Training Data
gbtModel = gbt.fit(trainingData)

# Use test set here so we can measure the accuracy of our model on new data
prediction_gbt = gbtModel.transform(testData)

# cvModel uses the best model found from the Cross Validation
# Evaluate best model
print("The area under ROC for test set with gradient boosting tree before CV is {}".format(evaluator.evaluate(prediction_gbt)))

The area under ROC for test set with gradient boosting tree before CV is 0.9383307386285076

#### Parameter Tuning and K-fold cross-validation

In [38]:
gbt = GBTClassifier(labelCol="label", featuresCol="features")

paramGrid = (ParamGridBuilder()
             .addGrid(gbt.maxDepth, [2, 4, 6])
             .addGrid(gbt.maxBins, [20, 60])
             .addGrid(gbt.maxIter, [10, 20])
             .build())

# Create 5-fold CrossValidator
cv = CrossValidator(estimator = gbt, estimatorParamMaps = paramGrid, evaluator = evaluator, numFolds = 5)

# Run cross validations.  This can take about 6 minutes since it is training over 20 trees!
cvModel = cv.fit(trainingData)

# Use test set here so we can measure the accuracy of our model on new data
prediction_gbt_best = cvModel.transform(testData)

# cvModel uses the best model found from the Cross Validation
# Evaluate best model
print("The area under ROC for test set with gradient boosting tree after CV is {}".format(evaluator.evaluate(prediction_gbt_best)))

The area under ROC for test set with gradient boosting tree after CV is 0.9524835828494627

## Get the best model with best hyper-parameter

In [40]:
import pandas as pd
roc_result = pd.DataFrame({"Logistic Regression": [0.9463, 0.9515], "Random Forest": [0.9252, 0.9253], "Gradient Boost": [0.9383, 0.9525], "Index": ["Before CV", "After CV"]})
roc_result.set_index("Index")

Out[ 49 ]: 
 Gradient Boost Logistic Regression Random Forest
Index 
Before CV 0.9383 0.9463 0.9252
After CV 0.9525 0.9515 0.9253

So based on the dataframe above, before cross validataion, logistic regression has the best performance in predicting the pet owner. But after cross validation, gradient boost tree has the best AUC and the prediction performance increases 1.5%. In general, I think these three models all did good job in prediction since the maximum of AUC is 1 and the AUC for these three models are pretty close to 1 even before the cross validation.

In [42]:
print("The best maximum depth is:", cvModel.bestModel._java_obj.getMaxDepth())
print("The best maximum bins is:", cvModel.bestModel._java_obj.getMaxBins())
print("The best maximum iteration is:", cvModel.bestModel._java_obj.getMaxIter())

The best maximum depth is: 6
The best maximum bins is: 20
The best maximum iteration is: 20

The best hyperparameter for the gradient boost tree is maximum depth = 6, maximum bins = 20, maximum iteration = 20 and with other parameters set to default.

## Apply the model

### Classify All The Users

Gathering the rest of unseen data from the original dataset.

In [47]:
df_clean_rest = df_clean_tmp1[1].union(df_clean_tmp0[1])

In [48]:
print("Rest Dataset Count: " + str(df_clean_rest.count()))
df_clean_rest.show(5)
#print("Rest Dataset with Label 1 Count: " + str(df_clean_rest.filter(col('label') == 1).count()))
#print("Rest Dataset with Label 0 Count: " + str(df_clean_rest.filter(col('label') == 0).count()))

Rest Dataset Count: 5762079
+--------------+---------+--------------------+
 creator_name| userid| comment|
+--------------+---------+--------------------+
 Aaron Rift|2212898.0|I need two of the...|
Aarons Animals| 426857.0|hey Aron I made a...|
Aarons Animals| 785631.0|I love the show m...|
Aarons Animals| 881725.0|my cat died like ...|
Aarons Animals|1351699.0|lol haha! This is...|
+--------------+---------+--------------------+
only showing top 5 rows

Building a new pipeline that includes the best model

In [50]:
from pyspark.ml.feature import RegexTokenizer, Word2Vec
from pyspark.ml import Pipeline

#Create final pipeline
regexTokenizer = RegexTokenizer(inputCol="comment", outputCol="words", pattern="\\W")
word2Vec = Word2Vec(inputCol="words", outputCol="features")
gbt_final = GBTClassifier(labelCol="label", featuresCol="features", maxDepth = 6, maxBins = 20, maxIter = 20)

pipeline_final = Pipeline(stages=[regexTokenizer, word2Vec, gbt_final])

# Fit the pipeline to train+test to get final model.
pipelineFit_final = pipeline_final.fit(df_clean_sub)
prediction_final = pipelineFit_final.transform(df_clean_rest)

The number of predicted Dog/Cat Owner is 564468, which is 22% of the total user. Compared with using only the key word to classify pet owner, the percentage is only 0.4% of the total user.

In [52]:
print("Total User: " + str(prediction_final.select('userid').distinct().count()))
print("Predicted Number of Dog/Cat Owner: " + str(prediction_final.filter(col('prediction') == 1).select('userid').distinct().count()))
print("Predicted Number of Non-Dog/Cat Owner: " + str(prediction_final.filter(col('prediction') == 0).select('userid').distinct().count()))

Total User: 2519673
Predicted Number of Dog/Cat Owner: 564468
Predicted Number of Non-Dog/Cat Owner: 2267445

In [53]:
print("Rest Dataset with label 1 Count: " + str(prediction_final.filter(col('label') == 1).select('userid').distinct().count()))
print("Rest Dataset with label 0 Count: " + str(prediction_final.filter(col('label') == 0).select('userid').distinct().count()))

Rest Dataset with label 1 Count: 11236
Rest Dataset with label 0 Count: 2515893

#### 3. Get insigts of Users

In [55]:
display(prediction_final.filter(col('prediction') == 1).select('comment', 'prediction'))

comment prediction I need two of these Donald pet toys one of for my cat and one for my dog. 1.0 hey Aron I made a video with my cat! You inspired me !!! I am a big fan of you! Do you have any tips for me? thank you! 1.0 I love the show my cat from hell #mycatfromhell 1.0 my cat died like if you agree 1.0 lol haha! This is so well made! I showed my cat loving cousins this and they died of laughter later on there channel is gonna be “Babycat Vlogs” if you can sub to them it would be greatly appreciated. The channel isn’t yet made but that’s what it will be called. 1.0 I did the same thing on my catI caught her sneaking out and buying drugs off a junkie in the alleyway then she smoked weed and also became the leader of the cat mafia 1.0 Ur cat is savage when i even try to put on a shirt with my cat i will become pirate hook 1.0 So a broad-headed skink got into my house and its hiding somewhere in my room I cant find it; any tips on how to lure it out of wherever its hiding? I dont want my dogs or cats to get it! 1.0 Lucy is so cute. she was looking at Bruce and Dexter to see if they were going to go. my dogs are like that too. lazy bums. hahaha 1.0 Is it normal for my dog to stick her snout on the back edge of the water bowl and lick the water up using the side of the bowl? 1.0 One of my dogs does that too so I have to give him a separate splash container... 😁 1.0 She looks bit like mine only hes a stud love watching you with her my dog is getting a treat God bless brother 1.0 Im so glad to see Anakin so healthy and beautiful at the age of 5! Cats seem to like doing things the hard way. One of my cats who passed away recently was frail and weak from hyperthyroidism but she ignored the kitty steps I put by my bed & the sofa & insisted on jumping/clawing her way up instead. You go Ani! 1.0 Omg my dogs name is bubu 1.0 id love to go for hikes with that pack! I wish there were fully fenced areas even 1/4 of that size where I could run my dogs. here all you have are partly fenced school fields and a few 1/2-1 acre fenced dog parks. if there was somewhere that safe I could run my dogs id have a pack of sighthounds in a minute. too many rabbits here too all over. have you ever tried vibration or training collars with any sight hounds? its common with bird dogs and scent hounds but I keep hearing sight hounds are too soft have you found this true? 1.0 Thanks I was watching this with my dog ( What was I thinking!!!!!!) Now he has dogmares about it. 1.0 The cat in the thumbnail looks like my cat 😂 1.0 Exactly how I talk to my cats: Hi baby!!! I also throw in Bushki Bubbeleh Boobie and Bubke. 1.0 I wish my cat was as playful as that 😹🙀 1.0 ah ive tried this so many time with my cats but they never seem to be interested lol 1.0 Hes so handsome...And its funny cause he acts just t like my cat at Home 1.0 Definitely going to try this with my cat! Shes fussy but I bet shed love this haha. :) 1.0 Wait a moment... We have a species of these here in South Carolina two of which are hanging out by my dog right now. I always thought that the Golden Orb Weavers here in the US were harmless... Are these the same species that have migrated via bananas in FL or are they a different one? Ive only seen a really big one when I was in GA. ...good thing I never tried to handle them as I didnt know they had venom comparable to Widows. 1.0 this guy can catch a wild lizard that can go very fast but my cat wont even stay in my room for a minute 1.0 oh! I think I found one of those a little while ago! so I was just playing soccer with my dog and then my dog went to go poop XD. you know how dogs kick the grass to mark their territory? well... while my dog was doing that he kicked something else too... I thought it was an earth worm but it was like one of those creatures 1.0 wow your amazing my cat would eat that spider 1.0 thanks noise triggered my dog 1.0 2:10 i look back and my cat was staring at me 1.0 Raptor fossil and its my cats birthday 1.0 The second most 

#### 4. Identify Creators With Cat And Dog Owners In The Audience

Brave Wilderness, Brian Barczyk, The Dodo and Taylor Nicole Dean are the creators with top 4 number of pet owners audience.

In [58]:
display(prediction_final.filter(col('prediction') == 1).select('creator_name').groupby('creator_name').count().orderBy('count', ascending=False))

creator_name count Brave Wilderness 150315 Brian Barczyk 70197 The Dodo 55844 Taylor Nicole Dean 52698 Robin Seplut 21836 Hope For Paws - Official Rescue Channel 21240 Vet Ranch 18538 Gohan The Husky 17077 Viktor Larkhill 15018 Think Like A Horse 14559 RaleighLink14 13041 ViralHog 12273 Cole & Marmalade 11822 Gone to the Snow Dogs 10596 Mr. Max T.V. 10482 Talking Kitty Cat 10288 stacyvlogs 9490 MonkeyBoo 9168 HammyLux 8120 Steff J 6717 Solid Gold Aquatics 6270 Zak Georges Dog Training rEvolution 6251 TobyTurner 5841 Paws Channel 5677 Aquarium Co-Op 4644 Happy Tails 4643 The Pet Collective 4412 Peter Caine Dog Training 4401 SlideShow ForFun 4395 null 4313 ViralBe 4180 Kitten Academy 4172 meow meow 3902 Lennon The Bunny 3874 Loki the Red Fox 3653 MARLENE MCCOHEN 3551 MaxluvsMya 3503 Big Cat Rescue 3388 Gotcha The Cockatoo 3318 LPE360 3260 Rachel Oleary 3119 GoHerping 3113 Pickles12807 3022 Cat Man Chris 3021 Dogumentary TV 2991 Exploring Alabama 2931 StormyRabbits 2676 Colorado Bigfoot 2286 Epic Wildlife 2258 Kamp Kenan 2195 Dan Hiteshew 2180 HoppingHammy 2126 Funny Pets 2125 prehistoricpetstv 2106 best pets 2001 PETA (People for the Ethical Treatment of Animals) 1993 FROSTY Life 1981 World Animal Awareness Society 1907 BrookIvy3 1877 Kyle Naegeli 1863 ERB2 1830 Creative Pet Keeping 1814 DarkDynastyK9s 1775 The Dro 1753 The Hoover Boys 1736 Gardea23 1731 Cockatoo Luck 1625 AQUAPROS 1557 CoralFish12g 1529 Max and Katie the Great Danes 1494 101rabbits 1478 Animal Adventure Park 1457 Daily Monkey 1453 Menthol Kratom 1431 REALITY TALK REVIEWS 1412 Gnat Vova 1339 foxalbiazul 1337 Emily Rios 1336 Shehry Vlogs 1332 My BB Bunny 1307 Cincinnati Zoo & Botanical Garden 1300 DanRock 1261 ChewyMinions AJ 1241 Teresa Bullock 1222 rsdfreetour 1222 Worlds Fuzziest Videos 1205 Kovu 1203 Sad Cat 1153 Einstein Parrot 1152 TheMeanKitty 1127 Feral Pride 1126 MAI PM 1103 Alex Knappenberger 1095 2CAN.TV - Ripley the Toucan! 1078 Top 10 COMPILATION 1075 Viral Maniacs 1027 LightningLpsTV 1006 eMusic Talent 989 Bulkreefsupply.com - Saltwater Aquarium Supplies 988 Rise Up Society Fan Page 966 Lil Bub 948 wskrsnwings 942 Olympus Reptiles 933 SnakeHuntersTV 926 viralvids 919 PugPibbleHedgie 918 AnimalBytesTV 914 Chriskillergaming 914 RedLeopardVideos 910 Schnauzer Mom 904 The Fatman 902 The Neighbors 900 Riot The Red Fox 883 The Humane Society of the United States 866 IFG 857 Shamama Hunting Owl 855 LOST HIGHWAY 843 carolinabullyfarms 837 Cute Cats Kwazi and Uli 828 Flip Aquatics - LupDiesel 816 Maymo 811 Jackson Galaxy 810 Inappropriate Reefer 809 Smarter Every Day 2 805 Aarons Animals 798 Reptile Channel 789 The Truth in Story 787 Garrett Rose 786 The Rat Guru 779 Frozen Kitten 777 teardropface 760 ShowMe 757 pastorbobncc 748 CatAleah 742 Deadly Tarantula Girl 736 Butters 724 Amethyst Is Cutie 721 pecktec 710 HitsTownUSA 700 Live2Thrill 695 Tom Moran 694 Louie The Beagle 688 unicornoxtreme GT 687 Thor Unleashed 685 Ginger Kitties Four 684 MyBackyardBirding 677 AfroHerpkeeper 667 Doug The Pug 665 76Highboy Reloading 664 Rocky Kanaka 664 Hooks Hollands - Ohio Holland Lops 655 Royal Kennels 648 Lacey Shih Tzu 639 wingsNpaws 629 Alex Dauterive 614 camelsandfriends 612 Maddie Smith 610 Top-5 Top-10 609 Floppycats.com 607 Kitsune Kreations 596 St. Bernard Acres 590 Donnie MFNG 586 DailyBigCat 585 MarineDepot.com 581 Dino Master 567 Rahbinah Rastaban 560 CJS AQUARIUMS 556 Edward Tyndall 553 Nora Oscar & Jack 551 Livin Surfside 546 mocha5cougar 544 SummerRose14 543 DWSDARIUS FISH TANKS 532 Pawfessor 525 NeonVisions 520 Connor OBrien 519 PENNSYLVANIA BIGFOOT & DOGMAN RESEARCH 515 MrPickles Channel 513 Rotter Tube Reef 512 Ericas Slot World 511 JOHN 3:16 Video Ministry 511 Balloon Animals 506 Animals TV 504 278pikelk 501 ojatro 491 StormyStrikesChannel 488 Mr. X Dreams 480 YamaChan 466 weedcat 454 bodhispeak 440 Steve Poland Aquatics 436 Snake Channel 434 OTIS FUNKMEYER 432 IKM218 431 RockStar FAIL 424 MileHighReefers 420 FishOfHex 417 Abbey Normal 417 The Budget E

The percentage of creators with cat and dog owners in the audience is 74.26%.

In [60]:
print("Total Number of Youtube Channel: " + str(prediction_final.select('creator_name').distinct().count()))

Out[ 35 ]: 3858

In [61]:
print("The Number of Youtube Channel with Dog/Cat Owner Comment: " + str(prediction_final.filter(col('prediction') == 1).select('creator_name').distinct().count()))

Out[ 36 ]: 2865

In [62]:
prediction_final.createOrReplaceTempView("pred_tmp")

In [63]:
%sql
select distinct p.creator_name, p1.sub_count, pt.total_count, p1.sub_count / pt.total_count as percentage
from pred_tmp as p inner join 
(select creator_name, count(distinct userid) as sub_count from pred_tmp where prediction = 1 group by 1) as p1 
on p.creator_name = p1.creator_name inner join
(select creator_name, count(distinct userid) as total_count from pred_tmp group by 1) as pt
on p.creator_name = pt.creator_name
where p.prediction = 1
order by sub_count desc;

creator_name sub_count total_count percentage Brave Wilderness 128004 633267 0.20213274969325734 Taylor Nicole Dean 41587 134537 0.309111991496763 The Dodo 39654 164977 0.240360777562933 Brian Barczyk 38834 135460 0.28668241547320245 Hope For Paws - Official Rescue Channel 18762 90601 0.20708380702199755 Vet Ranch 15698 65124 0.24104784718383393 Gohan The Husky 14563 63435 0.22957357925435484 Robin Seplut 12499 46051 0.27141647304075917 ViralHog 11278 70445 0.1600965292071829 Cole & Marmalade 9473 30751 0.308055022600891 Talking Kitty Cat 9085 48502 0.18731186342831224 stacyvlogs 8418 26001 0.323756778585439 Viktor Larkhill 8406 29796 0.2821184051550544 RaleighLink14 8242 17738 0.46465215920622394 Gone to the Snow Dogs 7416 25228 0.2939590930711907 Think Like A Horse 6590 14259 0.4621642471421558 MonkeyBoo 6425 37435 0.17163082676639507 Steff J 5522 19111 0.28894354036942077 Paws Channel 5373 22674 0.23696745170680075 Zak Georges Dog Training rEvolution 5309 12313 0.4311703078047592 Solid Gold Aquatics 4797 14206 0.33767422215965087 TobyTurner 4220 19435 0.21713403653202984 HammyLux 3877 8428 0.4600142382534409 The Pet Collective 3825 16538 0.2312855242471883 Happy Tails 3806 11354 0.335212259996477 ViralBe 3720 22037 0.16880700639833007 meow meow 3575 11834 0.30209565658272775 Mr. Max T.V. 3507 9721 0.36076535335870796 SlideShow ForFun 3330 12119 0.27477514646422974 LPE360 3031 17833 0.16996579375315427 Aquarium Co-Op 2935 10389 0.28251034748291465 Lennon The Bunny 2897 6443 0.4496352630762067 Gotcha The Cockatoo 2647 10906 0.24271043462314323 Big Cat Rescue 2602 11955 0.21764951902969468 Cat Man Chris 2474 8176 0.30259295499021527 Dogumentary TV 2397 7018 0.34155029923055 Pickles12807 2273 7137 0.3184811545467283 Loki the Red Fox 2265 10218 0.22166764533176747 Peter Caine Dog Training 2240 8025 0.2791277258566978 GoHerping 2193 5905 0.37138018628281116 StormyRabbits 2085 4251 0.49047282992237123 Epic Wildlife 2051 14738 0.13916406568055367 MARLENE MCCOHEN 2046 5049 0.40522875816993464 Funny Pets 2033 10706 0.18989351765365214 Rachel Oleary 1965 5902 0.33293798712300915 best pets 1937 6858 0.28244386118401865 Kitten Academy 1831 5951 0.307679381616535 ERB2 1815 14562 0.12463947259991759 Kyle Naegeli 1770 15045 0.11764705882352941 prehistoricpetstv 1751 10631 0.1647069889944502 PETA (People for the Ethical Treatment of Animals) 1742 9306 0.18719105953148507 World Animal Awareness Society 1671 6997 0.23881663570101472 HoppingHammy 1624 4553 0.35668789808917195 Kamp Kenan 1620 7313 0.22152331464515246 DarkDynastyK9s 1522 8862 0.17174452719476416 Gardea23 1501 7085 0.2118560338743825 CoralFish12g 1381 5453 0.2532550889418669 AQUAPROS 1365 6011 0.2270836799201464 Animal Adventure Park 1317 7057 0.18662321099617402 MaxluvsMya 1247 3111 0.4008357441337191 foxalbiazul 1245 8119 0.1533440078827442 Emily Rios 1244 5954 0.20893516963385958 REALITY TALK REVIEWS 1214 5298 0.22914307285768215 DanRock 1210 5370 0.22532588454376165 My BB Bunny 1202 6589 0.18242525421156472 Worlds Fuzziest Videos 1177 5190 0.22678227360308284 101rabbits 1159 2977 0.3893181054753107 The Dro 1150 4066 0.2828332513526808 Creative Pet Keeping 1138 3331 0.34163914740318224 Sad Cat 1132 4450 0.2543820224719101 MAI PM 1085 4100 0.2646341463414634 Alex Knappenberger 1046 4442 0.2354795137325529 Top 10 COMPILATION 1045 5332 0.19598649662415604 The Hoover Boys 1034 4518 0.2288623284639221 TheMeanKitty 1013 3805 0.2662286465177398 rsdfreetour 1012 4369 0.2316319523918517 Viral Maniacs 1007 14086 0.07148942212125514 Cincinnati Zoo & Botanical Garden 1001 5050 0.19821782178217823 2CAN.TV - Ripley the Toucan! 972 5070 0.19171597633136095 Colorado Bigfoot 970 3073 0.31565245688252525 Exploring Alabama 958 3249 0.2948599569098184 eMusic Talent 953 5296 0.17994712990936557 Feral Pride 939 4228 0.22209082308420056 viralvids 907 5723 0.15848331294775467 Shehry Vlogs 904 7292 0.12397147558968732 Dan Hiteshew 889 2219 0.40063091482649843 LightningLpsTV 860 2659 0.3234298608499436

In [64]:
%sql
select distinct p.creator_name, p1.sub_count, pt.total_count, p1.sub_count / pt.total_count as percentage
from pred_tmp as p inner join 
(select creator_name, count(distinct userid) as sub_count from pred_tmp where prediction = 1 group by 1) as p1 
on p.creator_name = p1.creator_name inner join
(select creator_name, count(distinct userid) as total_count from pred_tmp group by 1) as pt
on p.creator_name = pt.creator_name
where p.prediction = 1 and pt.total_count > 1000
order by sub_count desc;

creator_name sub_count total_count percentage Brave Wilderness 128004 633267 0.20213274969325734 Taylor Nicole Dean 41587 134537 0.309111991496763 The Dodo 39654 164977 0.240360777562933 Brian Barczyk 38834 135460 0.28668241547320245 Hope For Paws - Official Rescue Channel 18762 90601 0.20708380702199755 Vet Ranch 15698 65124 0.24104784718383393 Gohan The Husky 14563 63435 0.22957357925435484 Robin Seplut 12499 46051 0.27141647304075917 ViralHog 11278 70445 0.1600965292071829 Cole & Marmalade 9473 30751 0.308055022600891 Talking Kitty Cat 9085 48502 0.18731186342831224 stacyvlogs 8418 26001 0.323756778585439 Viktor Larkhill 8406 29796 0.2821184051550544 RaleighLink14 8242 17738 0.46465215920622394 Gone to the Snow Dogs 7416 25228 0.2939590930711907 Think Like A Horse 6590 14259 0.4621642471421558 MonkeyBoo 6425 37435 0.17163082676639507 Steff J 5522 19111 0.28894354036942077 Paws Channel 5373 22674 0.23696745170680075 Zak Georges Dog Training rEvolution 5309 12313 0.4311703078047592 Solid Gold Aquatics 4797 14206 0.33767422215965087 TobyTurner 4220 19435 0.21713403653202984 HammyLux 3877 8428 0.4600142382534409 The Pet Collective 3825 16538 0.2312855242471883 Happy Tails 3806 11354 0.335212259996477 ViralBe 3720 22037 0.16880700639833007 meow meow 3575 11834 0.30209565658272775 Mr. Max T.V. 3507 9721 0.36076535335870796 SlideShow ForFun 3330 12119 0.27477514646422974 LPE360 3031 17833 0.16996579375315427 Aquarium Co-Op 2935 10389 0.28251034748291465 Lennon The Bunny 2897 6443 0.4496352630762067 Gotcha The Cockatoo 2647 10906 0.24271043462314323 Big Cat Rescue 2602 11955 0.21764951902969468 Cat Man Chris 2474 8176 0.30259295499021527 Dogumentary TV 2397 7018 0.34155029923055 Pickles12807 2273 7137 0.3184811545467283 Loki the Red Fox 2265 10218 0.22166764533176747 Peter Caine Dog Training 2240 8025 0.2791277258566978 GoHerping 2193 5905 0.37138018628281116 StormyRabbits 2085 4251 0.49047282992237123 Epic Wildlife 2051 14738 0.13916406568055367 MARLENE MCCOHEN 2046 5049 0.40522875816993464 Funny Pets 2033 10706 0.18989351765365214 Rachel Oleary 1965 5902 0.33293798712300915 best pets 1937 6858 0.28244386118401865 Kitten Academy 1831 5951 0.307679381616535 ERB2 1815 14562 0.12463947259991759 Kyle Naegeli 1770 15045 0.11764705882352941 prehistoricpetstv 1751 10631 0.1647069889944502 PETA (People for the Ethical Treatment of Animals) 1742 9306 0.18719105953148507 World Animal Awareness Society 1671 6997 0.23881663570101472 HoppingHammy 1624 4553 0.35668789808917195 Kamp Kenan 1620 7313 0.22152331464515246 DarkDynastyK9s 1522 8862 0.17174452719476416 Gardea23 1501 7085 0.2118560338743825 CoralFish12g 1381 5453 0.2532550889418669 AQUAPROS 1365 6011 0.2270836799201464 Animal Adventure Park 1317 7057 0.18662321099617402 MaxluvsMya 1247 3111 0.4008357441337191 foxalbiazul 1245 8119 0.1533440078827442 Emily Rios 1244 5954 0.20893516963385958 REALITY TALK REVIEWS 1214 5298 0.22914307285768215 DanRock 1210 5370 0.22532588454376165 My BB Bunny 1202 6589 0.18242525421156472 Worlds Fuzziest Videos 1177 5190 0.22678227360308284 101rabbits 1159 2977 0.3893181054753107 The Dro 1150 4066 0.2828332513526808 Creative Pet Keeping 1138 3331 0.34163914740318224 Sad Cat 1132 4450 0.2543820224719101 MAI PM 1085 4100 0.2646341463414634 Alex Knappenberger 1046 4442 0.2354795137325529 Top 10 COMPILATION 1045 5332 0.19598649662415604 The Hoover Boys 1034 4518 0.2288623284639221 TheMeanKitty 1013 3805 0.2662286465177398 rsdfreetour 1012 4369 0.2316319523918517 Viral Maniacs 1007 14086 0.07148942212125514 Cincinnati Zoo & Botanical Garden 1001 5050 0.19821782178217823 2CAN.TV - Ripley the Toucan! 972 5070 0.19171597633136095 Colorado Bigfoot 970 3073 0.31565245688252525 Exploring Alabama 958 3249 0.2948599569098184 eMusic Talent 953 5296 0.17994712990936557 Feral Pride 939 4228 0.22209082308420056 viralvids 907 5723 0.15848331294775467 Shehry Vlogs 904 7292 0.12397147558968732 Dan Hiteshew 889 2219 0.40063091482649843 LightningLpsTV 860 2659 0.3234298608499436

#### 5. Analysis and Future work

In this project, the dataset of user comments for youtube videos related to animals or pets with 5820035 observations is used to identify cat or dog owners by using mainly pyspark ml package deployed on Databricks.

In the data processiong part, the observations with null comments were dropped and the new dataset ended up with 5818984 observations. Then based on the comments, some key words like "my cat", "my dog" and "my puppy" was created to first label the observaitons with dog/cat owners as label 1 and non dog/cat owners as lebel 0. After labeling the dataset, the train set and test set were generated for the model fitting part. Notice that data is very unbalanced in terms of the labeling, so in order to have a more balanced data, I first splitted the data with only label 1 and label 0  to get approximately equal labelling and then combined them to do the future split for train set and test set. So, the data that I ended up being used has the following information in terms of the labelling and observation number:

Dataset Count: 56905
Training Dataset Count: 39958
Test Dataset Count: 16947
Training Dataset with Label 1 Count: 19910
Training Dataset with Label 0 Count: 20048
Test Dataset with Label 1 Count: 8378
Test Dataset with Label 1 Count: 8569

With the sample data ready, I did tokenization and vectorization on the data and then I tried logistic regression, random forest and gradient boosting tree these three models to classify the pet owners. For each of the models, I performed a 5-fold cross validations and using metric AUC to evaluate the performance of the classifiers. The result is shown below:

               Gradient Boost  Logistic Regression  Random Forest
    Index                                                        
    Before CV          0.9383               0.9463         0.9252
    After CV           0.9525               0.9515         0.9253

So based on the dataframe above, before cross validataion, logistic regression has the best performance in predicting the pet owner. But after cross validation, gradient boost tree has the best AUC and the prediction performance increases 1.5%. In general, I think these three models all did good job in prediction since the maximum of AUC is 1 and the AUC for these three models are pretty close to 1 even before the cross validation. 

The best hyperparameter for the gradient boost tree is maximum depth = 6, maximum bins = 20, maximum iteration = 20 and with other parameters set to default. 

After having the best model, I built a new pipeline that also includes the best model to fit the whole data that I used to do the model training and then transformed it on the rest of the unseen data which has 5762079 observations.

The number of predicted Dog/Cat Owner is 564468, which is 22% of the total user. Compared with using only the key word to classify pet owner, the percentage is only 0.4% of the total user.

In terms of the comments of pet/dog owners, by reading some of their comments, I found that most of the video they are watching are just entertaining and most of the audience expressed their feelings towards the video. So a sentimental analysis based on the comments can further be done (having hard time working this on spark since it is pretty slow…. But still trying…. ). 

In terms of the youtuber, the percentage of creators with cat and dog owners in the audience is 74.26% and Brave Wilderness, Brian Barczyk, The Dodo and Taylor Nicole Dean are the creators with top 4 number of pet owners audience and meanwhile, these 4 youtubers has a relatively high number of subscribers (over 1M). Since they have a very large amount of audience, so the chance of having the pet owners will be also higher compared to those youtubers with less audience. Thus, it is not surprising that the percentage of pet owners for these youtubers is around 20%~30%. I then restricted the total number of comment audience to be over 1000 (considered as actively youtubers). StormyRabbits, wingsNpaws, Menthol, Kratom, RaleighLink14 and Think Like A Horse has approximately 50% of their audience as pet owners.

Based on the analysis above, some recommendations can be made. 
1. Youtube Video Recommendation (You may like section): after performing the sentimental analysis on the comments, Youtube can know the preference of the audience in more detail since it’s describing in words.)So based on this information, the video recommendation can be more precise along with the play history for that user. 
2. Ads Recommendation: Classifying the audience with different labels can help Youtube better promote the ads. For example, in this case, if the audience is predicted as pet owner, then the audience can be more likely to click a pet suppliance related ads especially for those youtubers whose majority audience are pet owners. And the ads can even appear on the google search page or gmail page to increase the click rate.